python/pyarrow/_parquet.pyx

```diff
@@ -849,6 +849,18 @@ cdef class FileMetaData(_Weakrefable):
         cdef Buffer buffer = sink.getvalue()
         return _reconstruct_filemetadata, (buffer,)
 
+    def __hash__(self):
+        def flatten(obj):
```

Comment:
I think it would be simpler, and probably faster, to serialize the metadata using `self._metadata.WriteTo` and then hash the resulting buffer contents.

Reply:
Note that it can be as simple as taking the hash of a `bytes` copy:
```python
>>> hash(stream.getvalue().to_pybytes())
5574314894425565867
```

Reply:
But, even then, this approach will not be very performant while a hash is supposed to compute quickly:
```python
>>> md = pq.read_metadata("/home/antoine/arrow/dev/python/lineitem-snappy.pq")
>>> md
<pyarrow._parquet.FileMetaData object at 0x7f87befc19e0>
  created_by: parquet-cpp-arrow version 15.0.0-SNAPSHOT
  num_columns: 16
  num_rows: 2568534
  num_row_groups: 21
  format_version: 2.6
  serialized_size: 32440
>>> md.__reduce__()
(<cyfunction _reconstruct_filemetadata at 0x7f87c02fea80>,
 (<pyarrow.Buffer address=0x7f88082086c0 size=32440 is_cpu=True is_mutable=True>,))
>>> hash(md.__reduce__()[1][0].to_pybytes())
8779236492631648618
>>> %timeit hash(md.__reduce__()[1][0].to_pybytes())
93.6 µs ± 144 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

It will be far worse with this PR's approach, since generating the dict is extremely costly:
```python
>>> %timeit md.to_dict()
2.95 ms ± 29.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

(note: milliseconds, as opposed to microseconds above)

Reply:
As a general basis, a `__hash__` does not need to take all contents exhaustively into account if doing so is too expensive. It's ok to limit ourselves to a smaller, faster subset. For example only the schema:
```python
>>> %timeit hash(repr(md.schema))
6.59 µs ± 27.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

You can also convert to a Arrow schema:
```python
>>> %timeit md.schema.to_arrow_schema()
4.94 µs ± 8.25 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
````

Unfortunately, `pa.Schema.__hash__` is too slow currently:
```python
>>> s = md.schema.to_arrow_schema()
>>> %timeit hash(s)
12.7 µs ± 25.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

This is because `pa.Schema.__hash__` takes the [slow Python way](https://github.com/apache/arrow/blob/a6e577d031d20a1a7d3dd01536b9a77db5d1bff8/python/pyarrow/types.pxi#L2524-L2525), while the C++ Schema has a useful and fast `fingerprint` method. (same comment for `pa.Field` and `pa.DataType)

Of course you don't have to do all this here, but a `__hash__` function that takes milliseconds to execute is definitely not a good idea.

Reply:
Thanks for confirming my suspicion the current approach was abysmal. Just wasn't sure _what_ belongs in there. At least schema but seemed like filemetadata was a superset in some way of schema, so ought to include a few other things, thus added a few more attributes in the followup; please feel free to suggest any changes there. :) 

Timings I have for https://github.com/apache/arrow/pull/39781/commits/45059380531161759fe093281ff97988c9207bca are:
```python
In [9]: %timeit hash(meta)
727 ns ± 2.69 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [10]: %timeit hash(schema)
1.2 µs ± 144 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```